pkg/operation/botanist/component/extensions/dnsrecord/dnsrecord.go

```diff
@@ -180,6 +182,11 @@ func (c *dnsRecord) deploy(ctx context.Context, operation string) (extensionsv1a
 				// If the DNSRecord is not yet Succeeded, reconcile it again.
 				_ = mutateFn()
 			} else {
+				// TODO (voelzmo) remove this when all DNSRecord secrets have migrated to a prefixed version
+				// Check if we need to migrate the referenced secret
+				if !strings.HasPrefix(c.dnsRecord.Spec.SecretRef.Name, "dnsrecord-") {
+					c.dnsRecord.Spec.SecretRef.Name = fmt.Sprintf("dnsrecord-%s", c.dnsRecord.Spec.SecretRef.Name)
+				}
```

Comment:
I tested locally the migration but it didn't work:

```
{"error":"observed generation outdated (1/2)","level":"error","msg":"DNSRecord shoot--foo--bar/bar-owner did not get ready yet","operation":"reconcile","shoot":"garden-foo/bar","time":"2021-11-23T10:47:26+02:00"}
```

I think you have to also add the `gardener.cloud/operation=reconcile` annotation to the dnsrecord in this case to trigger its reconciliation.

Can you check whether something like will do the trick?

```suggestion
				if !strings.HasPrefix(c.dnsRecord.Spec.SecretRef.Name, "dnsrecord-") {
					c.dnsRecord.Spec.SecretRef.Name = fmt.Sprintf("dnsrecord-%s", c.dnsRecord.Spec.SecretRef.Name)

					metav1.SetMetaDataAnnotation(&c.dnsRecord.ObjectMeta, v1beta1constants.GardenerOperation, operation)
				}
```

Reply:
I would propose something slightly different:
```suggestion
				if c.dnsRecord.Spec.SecretRef.Name != c.secret.Name {
					metav1.SetMetaDataAnnotation(&c.dnsRecord.ObjectMeta, v1beta1constants.GardenerOperation, operation)
					c.dnsRecord.Spec.SecretRef.Name = c.secret.Name
				}
```
In essence, if the secret name is not what it should be, set it to the right one. We don't need to check prefixes etc. here.

Reply:
@stoyanr , I see the point in your proposal. Adapting the owner DNSRecord from `ReconcileOnce` to `ReconcileOnChange` would definitely help with dealing with such spec changes in the future.
I was rather wondering whether this requested change must be a blocker for this PR or is the current state "good enough" to deal with the issue and the requested change can be done as improvement in separate PR?
I am fine with the current state of the PR and with the hardcoded check for the secret prefix or for secret name (as you  suggested). WDYT, can the suggested improvement be implemented in separate PR or do you insist to happen in this PR?